database/service_db.go

```diff
@@ -64,52 +87,119 @@ func (d *LevelDBServiceDB) All() ([]*service.Service, error) {
 				logrus.WithField("service", decodeErr.ID).Warning(decodeErr.Error())
 				continue
 			}
+			iter.Release()
 			return nil, err
 		}
 		services = append(services, s)
 	}
 	iter.Release()
-	if err := iter.Error(); err != nil {
-		return nil, err
-	}
-
-	return services, nil
+	return services, iter.Error()
 }
 
 // Delete deletes service from database.
-func (d *LevelDBServiceDB) Delete(id string) error {
+func (d *LevelDBServiceDB) Delete(idOrAlias string) error {
+	id, alias, err := d.getIDAndAlias(idOrAlias)
+	if err != nil {
+		return err
+	}
+	// delete alias
+	if alias != "" {
+		if err := d.aliases.Delete([]byte(alias), nil); err != nil {
+			return err
+		}
+	}
+	// delete service
 	return d.db.Delete([]byte(id), nil)
 }
 
 // Get retrives service from database.
-func (d *LevelDBServiceDB) Get(id string) (*service.Service, error) {
+func (d *LevelDBServiceDB) Get(idOrAlias string) (*service.Service, error) {
+	id, _, err := d.getIDAndAlias(idOrAlias)
+	if err != nil {
+		return nil, err
+	}
 	b, err := d.db.Get([]byte(id), nil)
 	if err != nil {
-		if err == leveldb.ErrNotFound {
-			return nil, &ErrNotFound{ID: id}
-		}
 		return nil, err
 	}
-
-	return d.unmarshal(id, b)
+	return d.unmarshal(idOrAlias, b)
 }
 
 // Save stores service in database.
 func (d *LevelDBServiceDB) Save(s *service.Service) error {
+	// check service
 	if s.ID == "" {
-		return errors.New("database: can't save service without id")
+		return errCannotSaveWithoutID
 	}
+	// encode service
 	b, err := d.marshal(s)
 	if err != nil {
 		return err
 	}
-
-	return d.db.Put([]byte(s.ID), b, nil)
+	// save service
+	if err := d.db.Put([]byte(s.ID), b, nil); err != nil {
+		return err
+	}
+	// save alias if exist
+	if s.Alias != "" {
+		if err := d.aliases.Put([]byte(s.Alias), []byte(s.ID), nil); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 // Close closes database.
 func (d *LevelDBServiceDB) Close() error {
-	return d.db.Close()
+	if err := d.db.Close(); err != nil {
+		return err
+	}
+	return d.aliases.Close()
+}
+
+// getIDAndAlias returns and separates id from alias.
+// id cannot be empty if an error is returned.
```

Comment:
I cannot understand what's the meaning of this

Reply:
this: "id cannot be empty if an error is returned." ?
if there is no error, then ID is never empty